src/sciline/data_graph.py

```diff
@@ -0,0 +1,205 @@
+    for combination in itertools.product(*constraints):
+        yield dict(zip(type_vars, combination))
+
+
+T = TypeVar('T', bound='DataGraph')
+
+
+class DataGraph:
+    def __init__(self, providers: None | Iterable[ToProvider | Provider]) -> None:
+        self._cbgraph = cb.Graph(nx.DiGraph())
+        for provider in providers or []:
+            self.insert(provider)
+
+    @classmethod
+    def _from_cyclebane(cls: type[T], graph: cb.Graph) -> T:
+        out = cls([])
+        out._cbgraph = graph
+        return out
+
+    def copy(self: T) -> T:
+        return self._from_cyclebane(self._cbgraph.copy())
+
+    def __copy__(self: T) -> T:
+        return self.copy()
+
+    @property
+    def _graph(self) -> nx.DiGraph:
+        return self._cbgraph.graph
+
+    def _get_clean_node(self, key: Key) -> Any:
+        """Return node ready for setting value or provider."""
+        if key is NoneType:
+            raise ValueError('Key must not be None')
+        if key in self._graph:
+            self._graph.remove_edges_from(list(self._graph.in_edges(key)))
+            self._graph.nodes[key].pop('value', None)
+            self._graph.nodes[key].pop('provider', None)
+            self._graph.nodes[key].pop('reduce', None)
+        else:
+            self._graph.add_node(key)
+        return self._graph.nodes[key]
+
+    def insert(self, provider: Union[ToProvider, Provider], /) -> None:
+        """
+        Insert a callable into the graph that provides its return value.
+
+        Parameters
+        ----------
+        provider:
+            Either a callable that provides its return value. Its arguments
+            and return value must be annotated with type hints.
+            Or a ``Provider`` object that has been constructed from such a callable.
+        """
+        if not isinstance(provider, Provider):
+            provider = Provider.from_function(provider)
+        return_type = provider.deduce_key()
+        if typevars := _find_all_typevars(return_type):
+            for bound in _mapping_to_constrained(typevars):
+                self.insert(provider.bind_type_vars(bound))
+            return
+        # Trigger UnboundTypeVar error if any input typevars are not bound
+        provider = provider.bind_type_vars({})
+        self._get_clean_node(return_type)['provider'] = provider
+        for dep in provider.arg_spec.keys():
+            self._graph.add_edge(dep, return_type, key=dep)
+
+    def __setitem__(self, key: Key, value: DataGraph | Any) -> None:
+        """
+        Provide a concrete value for a type.
+
+        Parameters
+        ----------
+        key:
+            Type to provide a value for.
+        param:
+            Concrete value to provide.
+        """
+        # This is a questionable approach: Using MyGeneric[T] as a key will actually
+        # not pass mypy [valid-type] checks. What we do on our side is ok, but the
+        # calling code is not.
```

Comment:
Can you elaborate why this fails in mypy?

Does this mean that every insertion of a param needs to have a `type: ignore[valid-type]` comment?

Reply:
Apparently `MyGeneric[T]` is not a valid type. `MyGeneric[MyType]` is ok. So no, not every insertion needs a `type: ignore[valid-type]`, just the less common ones using a generic (not specialized).

Reply:
Would it be better to annotate `__setitem__` with something more lenient? Maybe even `Any`? I don't like having legitimate usages flagged as errors. That is, I would prefer false negatives over false positives.

Reply:
We should chat about that, as I have struggled quite a bit coming up with something sensible that is not just `Any` everywhere.